File: apps/backend/src/services/memory/embeddings.ts
export const MEMORY_EMBEDDING_DIMENSIONS = 1536
export const DEFAULT_MEMORY_EMBEDDING_MODEL = "text-embedding-3-small"

export function memoryEmbeddingModel(): string {
  return process.env["OPENAI_EMBEDDING_MODEL"] ?? DEFAULT_MEMORY_EMBEDDING_MODEL
}

// Best-effort: every failure degrades to [] so recall falls back to full-text/metadata ranking.
export async function embedMemoryText(input: string): Promise<number[]> {
  if (!input.trim()) return []
  const apiKey = process.env["OPENAI_API_KEY"]
  if (!apiKey) return []
  const baseUrl = (process.env["OPENAI_BASE_URL"] ?? "https://api.openai.com/v1").replace(
    /\/+$/,
    "",
  )
  const res = await fetch(`${baseUrl}/embeddings`, {
    method: "POST",
    headers: { Authorization: `Bearer ${apiKey}`, "Content-Type": "application/json" },
    body: JSON.stringify({ model: memoryEmbeddingModel(), input }),
  })
  if (!res.ok) return []
  const body = (await res.json()) as { data?: { embedding?: number[] }[] }
  const embedding = body.data?.[0]?.embedding
  return Array.isArray(embedding) && embedding.length === MEMORY_EMBEDDING_DIMENSIONS
    ? embedding
    : []
}

export function memoryVectorLiteral(values: number[]): string {
  return `[${values.map((v) => (Number.isFinite(v) ? v.toFixed(8) : "0")).join(",")}]`
}

File: apps/backend/src/yomi/services/memory/embeddings.py
"""Memory embeddings — best-effort call into the configured OpenAI endpoint.

Port of apps/backend/src/services/memory/embeddings.ts.
"""

from __future__ import annotations

import httpx

from yomi.conf import settings

MEMORY_EMBEDDING_DIMENSIONS = 1536
DEFAULT_MEMORY_EMBEDDING_MODEL = "text-embedding-3-small"

_embedding_timeout = httpx.Timeout(60.0)


def memory_embedding_model() -> str:
    return settings.openai_embedding_model or DEFAULT_MEMORY_EMBEDDING_MODEL


# Best-effort: every failure degrades to [] so recall falls back to full-text/metadata ranking.
async def embed_memory_text(input_text: str) -> list[float]:
    if not input_text.strip():
        return []
    api_key = settings.openai_api_key
    if not api_key:
        return []
    base_url = settings.openai_base_url.rstrip("/")

    async with httpx.AsyncClient(timeout=_embedding_timeout) as client:
        res = await client.post(
            f"{base_url}/embeddings",
            headers={"Authorization": f"Bearer {api_key}", "Content-Type": "application/json"},
            json={"model": memory_embedding_model(), "input": input_text},
        )
    if res.status_code != 200:
        return []
    body = res.json()
    embedding = body.get("data", [{}])[0].get("embedding")
    if not isinstance(embedding, list) or len(embedding) != MEMORY_EMBEDDING_DIMENSIONS:
        return []
    return [float(v) for v in embedding]


def memory_vector_literal(values: list[float]) -> str:
    parts = [f"{v:.8f}" if v == v and abs(v) != float("inf") else "0" for v in values]
    return "[" + ",".join(parts) + "]"